pkg/urbit/noun/imprison.c

```diff
@@ -3,228 +3,453 @@
+          const c3_y* b_y)
+{
+  //  strip trailing zeroes.
   //
-  while ( a_w && !b_d[a_w - 1] ) {
+  while ( a_w && !b_y[a_w - 1] ) {
     a_w--;
   }
 
-  //  Check for cat.
-  //
   if ( !a_w ) {
-    return 0;
+    return (u3_atom)0;
   }
-  else if ( (1 == a_w) && !(b_d[0] >> 31) ) {
-    return (c3_w)b_d[0];
+  else {
+    u3i_slab sab_u;
+
+    u3i_slab_bare(&sab_u, 3, a_w);
+
+    u3t_on(mal_o);
+    {
+      //  zero-initialize last word
+      //
+      sab_u.buf_w[sab_u.len_w - 1] = 0;
+      memcpy(sab_u.buf_y, b_y, a_w);
+    }
+    u3t_off(mal_o);
+
+    return u3i_slab_moot_bytes(&sab_u);
   }
+}
 
-  //  Allocate, fill, return.
+/* u3i_words(): Copy [a] words from [b] into an atom.
+*/
+u3_atom
+u3i_words(c3_w        a_w,
+          const c3_w* b_w)
+{
+  //  strip trailing zeroes.
   //
-  u3t_on(mal_o);
-  {
-    c3_w len_w = 2 * a_w;
+  while ( a_w && !b_w[a_w - 1] ) {
+    a_w--;
+  }
 
-    if ( !(b_d[a_w - 1] >> 32) ) {
-      len_w--;
-    }
+  if ( !a_w ) {
+    return (u3_atom)0;
+  }
+  else {
+    u3i_slab sab_u;
+    u3i_slab_bare(&sab_u, 5, a_w);
 
-    c3_w*     nov_w = u3a_walloc(len_w + c3_wiseof(u3a_atom));
-    u3a_atom* nov_u = (void*)nov_w;
+    u3t_on(mal_o);
+    memcpy(sab_u.buf_w, b_w, (size_t)4 * a_w);
+    u3t_off(mal_o);
 
-    nov_u->mug_w = 0;
-    nov_u->len_w = len_w;
+    return u3i_slab_moot(&sab_u);
+  }
+}
 
-    //  Fill the words.
-    //
-    {
-      c3_w i_w, x_w, max_w = a_w - 1;
-      c3_d i_d;
+/* u3i_chubs(): Copy [a] chubs from [b] into an atom.
+*/
+u3_atom
+u3i_chubs(c3_w        a_w,
+          const c3_d* b_d)
+{
+  //  strip trailing zeroes.
+  //
+  while ( a_w && !b_d[a_w - 1] ) {
+    a_w--;
+  }
 
-      for ( i_w = 0; i_w < max_w; i_w++ ) {
-        i_d = b_d[i_w];
-        x_w = 2 * i_w;
-        nov_u->buf_w[x_w] = i_d & 0xffffffffULL;
-        x_w++;
-        nov_u->buf_w[x_w] = i_d >> 32;
-      }
+  if ( !a_w ) {
+    return (u3_atom)0;
+  }
+  else if ( 1 == a_w ) {
+    return u3i_chub(b_d[0]);
+  }
+  else {
+    u3i_slab sab_u;
+    u3i_slab_bare(&sab_u, 6, a_w);
 
-      {
-        i_d = b_d[i_w];
-        x_w = 2 * i_w;
-        nov_u->buf_w[x_w] = i_d & 0xffffffffULL;
-        x_w++;
-      }
+    u3t_on(mal_o);
+    {
+      c3_w* buf_w = sab_u.buf_w;
+      c3_w    i_w;
+      c3_d    i_d;
 
-      if ( x_w < len_w ) {
-        nov_u->buf_w[x_w] = i_d >> 32;
+      for ( i_w = 0; i_w < a_w; i_w++ ) {
+        i_d = b_d[i_w];
+        *buf_w++ = i_d & 0xffffffffULL;
+        *buf_w++ = i_d >> 32;
       }
     }
+    u3t_off(mal_o);
 
-    pro = u3a_to_pug(u3a_outa(nov_w));
+    return u3i_slab_mint(&sab_u);
   }
-  u3t_off(mal_o);
-
-  return pro;
 }
 
 /* u3i_mp(): Copy the GMP integer [a] into an atom, and clear it.
 */
-u3_noun
+u3_atom
 u3i_mp(mpz_t a_mp)
 {
-  c3_w  pyg_w = mpz_size(a_mp) * ((sizeof(mp_limb_t)) / 4);
-  c3_w* buz_w = u3a_slab(4 * pyg_w);
+  size_t   siz_i = mpz_sizeinbase(a_mp, 2);
+  u3i_slab sab_u;
+  u3i_slab_init(&sab_u, 0, siz_i);
```

Comment:
NB: the old implementation not only allocated unconditionally, but it measured the gmp integer with very low resolution, and often over-allocated. On 64-bit systems (ie, the only systems we currently support), `mp_limb_t` is 64 bits, so measurements were calculated in 128-bit increments (`(number-of-limbs * 2) * 4)`).

Reply:
> On 64-bit systems (ie, the only systems we currently support)

I could swear urbit ran on 32bit, ARM, Raspberry Pi. Isn't the 2GB loom limit driven by pointers needing to fit in 32 bit registers?

Reply:
The ARM/Raspi stuff is all 64-bit. As far as I know, urbit hasn't worked on 32-bit architectures as long as I've been around. There's no reason it can't, the loom limitations are orthogonal. It just hasn't been a priority.